File: raposeitor/raposeitor.cpp
#include <iostream>
#include <vector>
#include <string>
#include <string.h>

using namespace std;

vector<vector<string> > prog;
int r[2048];
int t[2048];

int *getend(string &s) {
    char letra; int num;
    sscanf(s.c_str(),"%%%c%d", &letra, &num);
    if (letra=='r') return &r[num];
    else return &t[num];
}

int getval(string &s) {
    if (s[0]=='%') return *(getend(s));
    else {
        int num;
        sscanf(s.c_str(),"%d", &num);
        return num;
    }
}

int getposrot(string &r) {
    for (int i=0;i<(int)prog.size();i++)
        if (prog[i][0]=="label" and prog[i][1]==r)
            return i;
    return prog.size();
}

int main(int argc, char *argv[]) {

    if (argc != 2) {
        cout << "Uso: " << argv[0] << " [arquivo.rap]" << endl;
        return 1;
    }

    FILE *f;
    if (!strcmp(argv[1],"-"))
        f = stdin;
    else
        f = fopen(argv[1],"r");
    if (f==NULL) {
        cout << "Erro ao abrir arquivo." << endl;
        return 1;
    }

    // carrega o programa
    char linha[2048];
    while (fgets(linha,2048,f)!=NULL) {
        vector<string> instr;
        char *c = strtok(linha," \r\n");
        if (c != NULL and !strcmp(c,"printf")) {
            string s = c;
            instr.push_back(s);
            c[strlen(c)]=' ';
            int ini, fim;
            for (int i=0;linha[i];i++)
                if (linha[i]=='"') {
                    ini=i;
                    break;
                }
            for (int i=ini+1;linha[i];i++)
                if (linha[i]=='"' and linha[i-1]!='\\') {
                    fim=i;
                    break;
                }
            s = "";
            for (int i=ini+1;i<fim;i++)
                s = s + linha[i];
            instr.push_back(s);
        } else {
            while (c != NULL) {
                if (c[0]==';') break; // comentario
                string s = c;
                instr.push_back(s);
                c = strtok(NULL, " \r\n");
            }
        }
        while (instr.size() < 4) instr.push_back("nop"); // filler
        prog.push_back(instr);

    }

    if (f != stdin) fclose(f);

    // executa o programa
    int pc = 0;
    while (pc < (int)prog.size()) {
        int *dest = getend(prog[pc][1]);
        int a = getval(prog[pc][2]);
        int b = getval(prog[pc][3]);
        if (prog[pc][0] == "mov") {
            *dest = a;
            pc++;
        } else if (prog[pc][0] == "read") {
            cin >> *dest;
            pc++;
        } else if (prog[pc][0] == "label") {
            pc++; // ignora
        } else if (prog[pc][0] == "less") {
            *dest = (a < b ? 1 : 0);
            pc++;
        } else if (prog[pc][0] == "greater") {
            *dest = (a > b ? 1 : 0);
            pc++;
        } else if (prog[pc][0] == "lesseq") {
            *dest = (a <= b ? 1 : 0);
            pc++;
        } else if (prog[pc][0] == "greatereq") {
            *dest = (a >= b ? 1 : 0);
            pc++;
        } else if (prog[pc][0] == "equal") {
            *dest = (a == b ? 1 : 0);
            pc++;
        } else if (prog[pc][0] == "diff") {
            *dest = (a != b ? 1 : 0);
            pc++;
        } else if (prog[pc][0] == "jf") {
            int val = getval(prog[pc][1]);
            if (val == 0) pc = getposrot(prog[pc][2]);
            else pc++;
        } else if (prog[pc][0] == "jt") {
            int val = getval(prog[pc][1]);
            if (val != 0) pc = getposrot(prog[pc][2]);
            else pc++;
        } else if (prog[pc][0] == "printf") {
            string s = prog[pc][1];
            for (int i=0;s[i];i++)
                if (s[i]=='\\' and s[i+1]=='n') {
                    cout << endl;
                    i++;
                } else
                    cout << s[i];
            pc++;
        } else if (prog[pc][0] == "printv") {
            cout << getval(prog[pc][1]);
            pc++;
        } else if (prog[pc][0] == "add") {
            *dest = a+b;
            pc++;
        } else if (prog[pc][0] == "sub") {
            *dest = a-b;
            pc++;
        } else if (prog[pc][0] == "mult") {
            *dest = a*b;
            pc++;
        } else if (prog[pc][0] == "div") {
            *dest = a/b;
            pc++;
        } else if (prog[pc][0] == "mod") {
            *dest = a%b;
            pc++;
        } else if (prog[pc][0] == "jump") {
            pc = getposrot(prog[pc][1]);
        } else if (prog[pc][0] == "and") {
            *dest = ( ((a != 0) and (b != 0)) ? 1 : 0 );
            pc++;
        } else if (prog[pc][0] == "or") {
            *dest = ( ((a != 0) or (b != 0)) ? 1 : 0 );
            pc++;
        } else if (prog[pc][0] == "not") {
            *dest = ( (a==0) ? 1 : 0 );
            pc++;
        } else if (prog[pc][0] == "store" or prog[pc][0] == "load") {
            int val = getval(prog[pc][1]);
            int base=0, desl=getval(prog[pc][2]);
            string aux = prog[pc][2];
            for (int i=0;i<(int)aux.size();i++)
                if (aux[i]=='(') {
                    string tt = aux.substr(i+1, aux.size());
                    base = getval(tt);
                    break;
                }
            if (prog[pc][0] == "store")
                r[base+desl] = val;
            else
                *dest = r[base+desl];
            pc++;
        } else if (prog[pc][0] == "nop") {
            pc++;
        } else {
            cout << "Instrucao invalida: " << prog[pc][0] << endl;
            break;
        }
    }

    return 0;
}

File: raposeitor/raposeitor.py
import sys

if len(sys.argv) != 2:
    print("Uso: $ python3", sys.argv[0], "[programa.rap]");
    exit()

f = open(sys.argv[1], 'r')
prog = f.read().split("\n")
f.close()

r = [-1 for i in range(2048)]
t = [-2 for i in range(2048)]

def getval(op):
    if op[0]=='%' and op[1]=='r':
        return r[ int(op.split("r")[1]) ]
    elif op[0]=='%' and op[1]=='t':
        return t[ int(op.split("t")[1]) ]
    else:
        return int(op)

def atribui(rdest, val):
    if rdest[0]=='%' and rdest[1]=='r':
        r[ int(rdest.split("r")[1]) ] = val
    elif rdest[0]=='%' and rdest[1]=='t':
        t[ int(rdest.split("t")[1]) ] = val
    else:
        print("ERRO ERRO ERRO ERRO!!!11!!")

def poslabel(rotulo):
    i = 0
    while i < len(prog):
        if prog[i].split(" ")[0] == "label" and prog[i].split(" ")[1] == rotulo:
            return i
        i = i + 1
    return -1 # erro

def equal(rdest, op1, op2):
    if getval(op1) == getval(op2):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def diff(rdest, op1, op2):
    if getval(op1) != getval(op2):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def less(rdest, op1, op2):
    if getval(op1) < getval(op2):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def lesseq(rdest, op1, op2):
    if getval(op1) <= getval(op2):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def greater(rdest, op1, op2):
    if getval(op1) > getval(op2):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def greatereq(rdest, op1, op2):
    if getval(op1) >= getval(op2):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def nega(rdest, op):
    if (getval(op) == 0):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def ora(rdest, op1, op2):
    if (getval(op1) != 0 or getval(op2) != 0):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def anda(rdest, op1, op2):
    if (getval(op1) != 0 and getval(op2) != 0):
        atribui(rdest, 1)
    else:
        atribui(rdest, 0)

def imprimir(s):
    ini=0
    fim=0
    for i in range(len(s)):
        if s[i]=='"':
            ini=i+1
            break
    for j in range(ini, len(s)):
        if (s[j]=='"' and s[j-1]!='\\'):
            fim=j-1
            break
    i = ini
    while i <= fim:
        if (s[i]=='\\' and s[i+1]=='n'):
            print()
            i = i + 2
        elif (s[i]=='\\' and s[i+1]=='"'):
            print("\"", end="")
            i = i + 2
        else:
            print(s[i], end="")
            i = i + 1

pc = 0
while pc < len(prog):
    linha = prog[pc].strip().split(" ")
    while len(linha) < 4:
        linha.append("nop")
    instr = linha[0]
    rdest = linha[1].split(",")[0]
    op1 = linha[2].split(",")[0]
    op2 = linha[3].split(",")[0]

    if instr == ";" or instr == "":
        pc = pc + 1 #ignora
    elif instr == "printf":
        imprimir(prog[pc].strip());
        pc = pc + 1
    elif instr == "printv":
        print(getval(rdest),end="")
        pc = pc + 1
    elif instr == "read":
        atribui(rdest, int(input()))
        pc = pc + 1
    elif instr == "mov":
        atribui(rdest, getval(op1))
        pc = pc + 1
    elif instr == "add":
        atribui(rdest, getval(op1) + getval(op2))
        pc = pc + 1
    elif instr == "sub":
        atribui(rdest, getval(op1) - getval(op2))
        pc = pc + 1
    elif instr == "mult":
        atribui(rdest, getval(op1) * getval(op2))
        pc = pc + 1
    elif instr == "div":
        atribui(rdest, getval(op1) // getval(op2))
        pc = pc + 1
    elif instr == "mod":
        atribui(rdest, getval(op1) % getval(op2))
        pc = pc + 1
    elif instr == "not":
        nega(rdest, op1)
        pc = pc + 1
    elif instr == "or":
        ora(rdest, op1, op2)
        pc = pc + 1
    elif instr == "and":
        anda(rdest, op1, op2)
        pc = pc + 1
    elif instr == "equal":
        equal(rdest, op1, op2)
        pc = pc + 1
    elif instr == "diff":
        diff(rdest, op1, op2)
        pc = pc + 1
    elif instr == "less":
        less(rdest, op1, op2)
        pc = pc + 1
    elif instr == "lesseq":
        lesseq(rdest, op1, op2)
        pc = pc + 1
    elif instr == "greater":
        greater(rdest, op1, op2)
        pc = pc + 1
    elif instr == "greatereq":
        greatereq(rdest, op1, op2)
        pc = pc + 1
    elif instr == "load":
        atribui(rdest, r[ getval(op1.split("(")[0]) + getval(op1.split("(")[1].split(")")[0]) ])
        pc = pc + 1
    elif instr == "store":
        r[ getval(op1.split("(")[0]) + getval(op1.split("(")[1].split(")")[0]) ] = getval(rdest)
        pc = pc + 1
    elif instr == "label":
        pc = pc + 1
    elif instr == "jf":
        if getval(rdest) == 0:
            pc = poslabel(op1)
        else:
            pc = pc + 1
    elif instr == "jt":
        if getval(rdest) != 0:
            pc = poslabel(op1)
        else:
            pc = pc + 1
    elif instr == "jump":
        pc = poslabel(rdest)
    else:
        break
